crates/red_knot_python_semantic/src/types.rs

```diff
@@ -721,6 +721,45 @@ impl<'db> Type<'db> {
             Type::Tuple(_) => builtins_symbol_ty(db, "tuple"),
         }
     }
+
+    /// Return the string representation of this type when converted to string as it would be
+    /// provided by the `__str__` method. If that can't be determined, return `None`.
+    ///
+    /// When not available, this should fall back to the value of `[Type::repr]`.
+    /// Note: this method is used in the builtins `format`, `print`, `str.format` and `f-strings`.
+    pub fn str(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let str_result = match self {
+            Type::IntLiteral(_) => None,
+            Type::BooleanLiteral(_) => None,
+            Type::StringLiteral(_) => Some(*self),
+            // TODO: handle more complex types
+            _ => None,
+        };
+        str_result.or_else(|| self.repr(db))
+    }
+
+    /// Return the string representation of this type as it would be provided by the  `__repr__`
+    /// method at runtime. If that can't be determined, return `None`.
+    pub fn repr(&self, db: &'db dyn Db) -> Option<Type<'db>> {
```

Comment:
Same as above, I don't think we need to return an `Option` here; the "we don't know" fallback case can simply be `str`. That is, `builtins_symbol_ty(db, "str").to_instance(db)`

```suggestion
    /// method at runtime.
    pub fn repr(&self, db: &'db dyn Db) -> Type<'db> {
```